sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java

```diff
@@ -73,44 +73,65 @@ public class InformationSchema extends AbstractSchema
   private static final String SCHEMATA_TABLE = "SCHEMATA";
   private static final String TABLES_TABLE = "TABLES";
   private static final String COLUMNS_TABLE = "COLUMNS";
-  private static final RowSignature SCHEMATA_SIGNATURE = RowSignature
```

Comment:
nit: why the change from using `RowSignature`? Just thinking ahead and wondering if this has any impact on if we want to make `INFORMATION_SCHEMA` and `sys` tables be executed with the Druid engine (instead of the "bindable" fallback engine). If they are modeled as native datasources then we would want these to be defined in terms of the native types...

Reply:
oh i see in the description something about nullable bigint, but I think the problem is probably using `RowSignatures.toRelDataType` which only allows numeric columns to be nullable if `druid.generic.useDefaultValueForNull=false`, which is valid for native druid engine, but probably isn't the correct behavior for these tables.

I wonder if it would it be the correct behavior if these were executed by the Druid engine though and `druid.generic.useDefaultValueForNull=true`? I hope to switch the default to be `false` someday to be more sql compliant out of the box, but perhaps worth thinking about.

Reply:
This is, in fact, the core issue in this PR. If we use a Druid signature and round-trip to SQL, we lose SQL semantics. However, these are SQL, not Druid tables, so SQL semantics apply.

If we were to execute system tables queries with a Druid engine then the most obvious choice is to integrate SQL semantics into the Druid engine. That is, numeric types can be null. There may be historical reasons why numbers can't be null in Druid, but there are also good historical reasons why numeric types can be null in SQL.

The question is whether we want to alter SQL to follow Druid semantics or the other way around. I a consequence of using Druid semantics is that numbers sort in the wrong order, then I do suspect SQL is more right than Druid in this one case.

Reply:
>If we were to execute system tables queries with a Druid engine then the most obvious choice is to integrate SQL semantics into the Druid engine. That is, numeric types can be null. There may be historical reasons why numbers can't be null in Druid, but there are also good historical reasons why numeric types can be null in SQL.

I think you misunderstood or I wasn't very articulate; numeric types can already be nullable in Druids native type system, we just have a mode, `druid.generic.useDefaultValueForNull=true`, in which numbers are never null and replaced with zeros. This is embedded in the conversion logic to allow the planner to optimize things when it knows things will never be nullable. The problem here is using that same conversion logic for tables that don't follow the rule.

I guess the main thing I was trying to discuss here is that afaik there are still (non-immediate) plans to push all system tables down into the druid engine instead of using the fallback "bindable" engine, so that we have a consistent set of behavior and operators no matter what table you are querying, and at which point we would probably want these schemas to be expressed in Druid native type system for describing tables, (`RowSignature`), so I was wondering if that means we would just end up be changing these back at that point.

I think its probably fine to make this change, git remembers, so we can always fetch these `RowSignatures` (and modify to use nullable numerics) from git history if/when we make the system tables some sort of native druid table and queries in the future.

Reply:
@clintropolis, we  can make the needed change now. The thing you cited sounds like a Druid property, which means it applies to all columns? Or, is there a way to do the SQL thing and set the nullability bit on each column? That is, the other way to solve the issue is to say:

```java
RowSignature
      .builder()
      ...
      .add("ORDINAL_POSITION", ColumnType.BIGINT) // Non-nullable
      ...
      .add("CHARACTER_MAXIMUM_LENGTH", ColumnType.BIGINT, true) // Nullable
```

Do we have the ability to do something like the above today?